harness/determined/core/_profiler.py

```diff
@@ -0,0 +1,440 @@
+        except pynvml.NVMLError as ne:
+            logging.info(f"Error accessing NVML {ne}. GPU metrics will not be collected.")
+        except Exception as e:
+            logging.error(f"Error initializing pynvml: {e}")
+
+    def collect(self) -> typing.Dict[str, typing.Any]:
+        metrics = {}
+
+        for uuid, handle in self._pynvml_device_handles.items():
+            gpu_util = pynvml.nvmlDeviceGetUtilizationRates(handle).gpu
+            free_memory = pynvml.nvmlDeviceGetMemoryInfo(handle).free
+            metrics.update(
+                {
+                    uuid: {
+                        "gpu_util": gpu_util,
+                        "gpu_free_memory": free_memory,
+                    }
+                }
+            )
+        return metrics
+
+
+class _Metric:
+    def __init__(
+        self,
+        group: str,
+        metrics: typing.Dict[str, typing.Any],
+        timestamp: datetime.datetime,
+    ):
+        self.group = group
+        self.metrics = metrics
+        self.timestamp = timestamp
+
+
+class _Collector(threading.Thread):
+    def __init__(
+        self,
+        metrics_queue: queue.Queue,
+        collection_interval: int = 1,
+        max_collection_s: int = 60 * 30,
+    ):
+        self._collection_interval = collection_interval
+        self._metrics_queue = metrics_queue
+        self._max_collection_s = max_collection_s
+        self._metric_collectors = [
+            _GPU(),
+            _CPU(),
+            _Memory(),
+            _Disk(),
+            _Network(),
+        ]
+        self._should_exit = threading.Event()
+
+        super().__init__(daemon=True)
+
+    def run(self) -> None:
+        collection_stop_ts = time.time() + self._max_collection_s
+        while not self._should_exit.is_set():
+            if time.time() > collection_stop_ts:
+                # Max collection time reached. Notify queue consumers and exit.
+                self._metrics_queue.put(None)
+                return
+
+            next_collection_ts = time.time() + self._collection_interval
+            self._collect_metrics()
+            wait_ts = max(next_collection_ts - time.time(), 0)
+            self._should_exit.wait(timeout=wait_ts)
+
+    def stop(self) -> None:
+        self._should_exit.set()
+
+    def _collect_metrics(self) -> None:
+        for collector in self._metric_collectors:
+            timestamp = datetime.datetime.now(datetime.timezone.utc)
+            metrics = collector.collect()
+            if not metrics:
+                continue
+            group_metrics = _Metric(group=collector.group, metrics=metrics, timestamp=timestamp)
+            self._metrics_queue.put(group_metrics)
+
+
+class _Shipper(threading.Thread):
+    def __init__(
+        self,
+        session: api.Session,
+        agent_id: str,
+        trial_id: int,
+        run_id: int,
+        metrics_queue: queue.Queue,
+        flush_interval: int = 10,
```

Comment:
Similar question as with `_Collector.max_collection_s`. Not sure whether best practices is to hardcode the thing or not. So since I'm not sure, I'll just raise the issue and let you decide.

Reply:
Flush interval has been removed.